### PR TITLE
Add Count command

### PR DESCRIPTION
**Pull request by jpxd**
Sometimes you need to count db objects. To fetch all users in a list with `db.All(&users)` and use the list size is very memory inefficient when you have lots of entries.
I added the possibility for `db.Count(User{})` which counts all elements in that bucket without allocating a large list of Users.

### CODE REVIEW

File: count.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+func (n *Node) count(tx *bolt.Tx, info *modelInfo, count *int) error {
+	bucket := n.GetBucket(tx, info.Name)
+	if bucket == nil {
+		return fmt.Errorf("bucket %s not found", info.Name)
+	}
+
+	*count = 0
+	c := bucket.Cursor()
+	for k, v := c.First(); k != nil && v != nil; k, v = c.Next() {
+		(*count)++
```
Comment (asdine) [inline count.go]: The thing with your implementation is that you also count sub buckets and indexes.
Your tests works because the indexes are prefixed with `__` so they are put at the end of the bucket, but if i change the user name to "___john" it breaks.

In every bucket that uses indexes, Storm creates one sub bucket per index. So if you have 100 regular items in your bucket, in reality you have 100 + the number of indexes.
(example for User: it has one index for `Name`, one for `DateOfBirth` and one for `Slug` => 103 items)

When using a `Cursor`, if `Next` finds a sub bucket it returns a `nil` value so you can easily skip sub buckets by adding a simple `if` statement in you loop and changing your condition:

```go
for k, v := c.First(); k != nil; k, v = c.Next() {
 if v == nil {
   continue
 }
 (*count)++
}
```

**Comment by jpxd** [comment]
Thanks for pointing that out @asdine :)

File: count_test.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+	count, _ := db.Count(User{})
+	assert.Equal(t, 100, count)
+
+	w := User{Name: "John", ID: 101, Slug: fmt.Sprintf("John%d", 101), DateOfBirth: time.Now().Add(-time.Duration(101*10) * time.Minute)}
+	err := db.Save(&w)
+	assert.NoError(t, err)
+
+	count, _ = db.Count(User{})
+	assert.Equal(t, 101, count)
```
Comment (bep) [inline count_test.go]: The tests are missing a couple of branches.

Comment (jpxd) [inline count_test.go]: What do you mean?

Comment (bep) [inline count_test.go]: There is count on:

1) Node
2) DB
3) When in a transaction and not

Your tests cover only the first of the above.

File: count.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	bucket := n.GetBucket(tx, info.Name)
+	if bucket == nil {
+		return fmt.Errorf("bucket %s not found", info.Name)
+	}
+
+	*count = 0
+	c := bucket.Cursor()
+	for k, v := c.First(); k != nil; k, v = c.Next() {
+		if v == nil {
+			continue
```
Comment (bep) [inline count.go]: Also it would be nice with a test for the v == nil case.

**Comment by asdine** [comment]
@jpxd still interested?

**Comment by jpxd** [comment]
@asdine Yeah but I currently don't have the time to implement the additional tests wanted by @bep.
Feel free to either merge or close this PR...
